src/Nethermind/Nethermind.Era1/E2StoreReader.cs

```diff
@@ -0,0 +1,210 @@
+    {
+        _file = file;
+        _fileLength = RandomAccess.GetLength(_file);
+    }
+
+    public (T, long) ReadEntryAndDecode<T>(long position, Func<Memory<byte>, T> decoder, ushort expectedType)
+    {
+        Entry entry = ReadEntry(position, expectedType);
+
+        int length = (int)entry.Length;
+        using ArrayPoolList<byte> buffer = new ArrayPoolList<byte>(length, length);
+        RandomAccess.Read(_file, buffer.AsSpan(), position + HeaderSize);
+        return (decoder(buffer.AsMemory()), (long)(entry.Length + HeaderSize));
+    }
+
+    public async Task<(T, long)> ReadSnappyCompressedEntryAndDecode<T>(long position, Func<Memory<byte>, T> decoder, ushort expectedType, CancellationToken token = default)
+    {
+        Entry entry = ReadEntry(position, expectedType);
+        T returnValue = await ReadEntryValueAsSnappy(position + HeaderSize, entry.Length, decoder, token);
+        return ((T, long))(returnValue, entry.Length + HeaderSize);
+    }
+
+    public Entry ReadEntry(long position, ushort? expectedType, CancellationToken token = default)
+    {
+        ushort type = ReadUInt16(position);
+        uint length = ReadUInt32(position + 2);
+        ushort reserved = ReadUInt16(position + 6);
+
+        Entry entry = new Entry(type, length);
+        if (expectedType.HasValue && entry.Type != expectedType) throw new EraException($"Expected an entry of type {expectedType}, but got {entry.Type}.");
+        if (entry.Length + (ulong)position > (ulong)_fileLength)
+            throw new EraFormatException($"Entry has an invalid length of {entry.Length} at position {position}, which is longer than stream length of {_fileLength}.");
+        if (entry.Length > ValueSizeLimit)
+            throw new EraException($"Entry exceeds the maximum size limit of {ValueSizeLimit}. Entry is {entry.Length}.");
+        if (reserved != 0)
+            throw new EraFormatException($"Reserved header bytes has invalid values at position {position}.");
+        return entry;
+    }
+
+    private async Task<T> ReadEntryValueAsSnappy<T>(long offset, ulong length, Func<Memory<byte>, T> decoder, CancellationToken cancellation = default)
+    {
+        using ArrayPoolList<byte> inputBuffer = new ArrayPoolList<byte>((int)length, (int)length);
+        RandomAccess.Read(_file, inputBuffer.AsSpan(), offset);
+        Stream inputStream = inputBuffer.AsMemory().AsStream();
+
+        await using SnappyStream decompressor = new(inputStream, CompressionMode.Decompress, true);
+        await using RecyclableMemoryStream stream = RecyclableStream.GetStream(nameof(E2StoreReader));
+        await decompressor.CopyToAsync(stream, cancellation);
+
+        if (!stream.TryGetBuffer(out ArraySegment<byte> segment))
+        {
+            throw new InvalidDataException("Unable to get buffer for memory stream");
+        }
+
+        return decoder(segment);
+    }
+
+    public void Dispose()
+    {
+        _file.Dispose();
```

Comment:
So we dispose it even if it's passed externally? And we don't need to care about double disposal?